wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -223,6 +226,84 @@ newUnsignedTransaction ActiveWallet{..} options accountId payees = runExceptT $
+            -- Now we have to generate the change addresses needed,
+            -- because 'newUnsignedTransaction' function cannot do it by itself.
+            changeAddresses <- withExceptT NewTransactionErrorCreateAddressFailed $
+                genChangeOuts (utxChange unsignedTx)
+                              srcAccId
+                              spendingPassword
+                              (walletPassive activeWallet)
+            -- We have to provide source addresses and derivation paths for this transaction.
+            -- It will be used by external party to provide a proof that it has a right to
+            -- spend this money.
+            let srcAddresses         = extractSrcAddressesFrom unsignedTx
+                srcHdAddresses       = NonEmpty.map (toHdAddress db) srcAddresses
+                derivationPaths      = buildDerivationPathsFor srcHdAddresses srcAccId
+                allOutputs           = (NonEmpty.toList $ utxOutputs unsignedTx) ++ changeAddresses
+                unsignedTxWithChange = unsignedTx { utxOutputs = NonEmpty.fromList allOutputs }
+                finalTx              = toRegularTx unsignedTxWithChange
+            return (finalTx, NonEmpty.zip srcAddresses derivationPaths)
+  where
+    -- Take addresses that correspond to inputs of transaction.
+    extractSrcAddressesFrom :: UnsignedTx -> NonEmpty Address
+    extractSrcAddressesFrom unsignedTx =
+        NonEmpty.map (\(_, outAux) -> txOutAddress . toaOut $ outAux) $ utxOwnedInputs unsignedTx
+
+    buildDerivationPathsFor :: NonEmpty HD.HdAddress
+                            -> HD.HdAccountId
+                            -> NonEmpty [Word32]
+    buildDerivationPathsFor srcAddresses srcAccountId =
+        flip NonEmpty.map srcAddresses $ \(HD.HdAddress srcAddressId _) ->
+            let HD.HdAccountId _ (HD.HdAccountIx srcAccIndex)  = srcAccountId
+                HD.HdAddressId _ (HD.HdAddressIx srcAddrIndex) = srcAddressId
+            in [srcAccIndex, srcAddrIndex]
+
+    toHdAddress :: DB
+                -> Address
+                -> HD.HdAddress
+    toHdAddress db srcAddress =
+        -- Sine 'srcAddress' is returned by 'newUnsignedTransaction'
+        -- we assume that this address is valid so we definitely have
+        -- corresponding 'HdAddress'.
+        either (error "Impossible happened: source address doesn't have corresponding HD-address!")
+               id
+               (lookupCardanoAddress db srcAddress)
+
+    -- Convert 'UnsignedTx' to 'Tx', this regular 'Tx'
+    -- will be signed later (technically - the hash of the
+    -- transaction will be signed).
+    toRegularTx :: UnsignedTx -> Tx
+    toRegularTx unsignedTx =
+        let inputs  = NonEmpty.map fst $ utxOwnedInputs unsignedTx
+            outputs = NonEmpty.map toaOut $ utxOutputs unsignedTx
+            -- Currently all transactions have default (actually - empty) attributes.
+            attribs = def :: TxAttributes
```

Comment:
I get the rationale behind this, but it also makes me _a little_ uneasy: what if this won't be true anymore in future? what if deep down in the code that creates a new unsigned transaction, now they come with attributes -- our code would silently break 😲 

One possibility you have here is to extend the `UnsignedTx` type to carry around an extra field called `utxAttributes` that you can compute in `mkStdUnsignedTx`. There, you would call `makeUnsignedAbstractTx` (from `Pos.Client.Txp.Util`) which can now be extended to return attributes in the future, without this code breaking.